### PR TITLE
fix: abort skipped write reservations instead of committing them

### PR DESCRIPTION
**Pull request by divyvasal**
## What

Adds `L1Manager.abort_write` / `StorageManager.abort_write` (release a write reservation without publishing it) and uses it at the three call sites that today either leak the reservation or publish never-written objects:

1. **`PickleTransferStrategy.commit_store`** (the bug that motivated this): reservations it cannot fill (missing chunk, `tensor is None`, shape mismatch) were silently `continue`d past and left **write-locked forever**. Under a sustained workload the leaked locks pin L1 until every subsequent `reserve_write` fails. Observed in production (engine_driven + pickle transport): 604/1276 L1 objects write-locked and 91.5% memory usage within ~30 minutes, with every store returning `False`. A `warning` log with per-reason skip counts replaces the silent skips.
2. **`ShmTransferStrategy.prepare_store`**: reserved keys that got no SHM slot were released via `finish_write`, which fires `on_l1_keys_write_finished`, so a store listener (e.g. the L2 StoreController) can then persist the **unwritten** contents, and lookups see a valid-looking empty object.
3. **Instance-teardown sweep** (`engine_driven_transfer`): stale pending writes from a dead worker were likewise committed via `finish_write`; the slots hold unwritten or partially written data.

`abort_write` unlocks + frees under the manager lock and notifies `on_l1_keys_deleted_by_manager` (never `write_finished`), so no listener can observe the discarded objects. New `SM_WRITE_ABORTED` event mirrors `SM_WRITE_FINISHED`.

## Tests

- `tests/v1/distributed/test_abort_write.py` (new, real `StorageManager`): lock release + re-reservability, no `write_finished` listener publication, wrong-state/missing-key errors, and a `commit_store` regression test (short chunk list → `False`, zero leaked locks).
- Updated the two existing release-semantics asserts (`prepare_store` unused keys, unregister sweep) from `finish_write` to `abort_write`.
- Full `test_engine_driven_transfer.py` + `test_report_status.py` + new file: **40 passed, 2 skipped** on CUDA (H/B-series box, python 3.10).

## Notes

Found while productionizing #3892; same deployment also surfaced that `--l1-use-lazy` (default) disables the SHM transport via `_compute_shm_pool_info`, which is what routes engine_driven workers onto this pickle path in the first place. Happy to file separately.

---

**Update (rebased on current dev).** Rebased across the ~350 commits since this was opened. Two conflicts, both from dev adding code at the same spots: dev added a `force` flag to `L1Manager.delete()` right below where `abort_write` goes (kept both), and dev added a `logger.error` at each skip branch of the pickle `commit_store`, where this PR had added skip-reason tallies. I kept dev's per-skip logs and dropped the tallies, since dev's "store incomplete" error already reports reserved versus written. The abort itself is unchanged.

All three sites are still live on dev: the pickle path still leaves skipped reservations write-locked, and the SHM `prepare_store` and the dead-worker reap still `finish_write` objects that were never written.

Verified on an A10G with torch 2.13.0+cu130 (the repo's build pin): `test_abort_write.py` passes 4/4, and reverting the pickle-path fix fails exactly `test_pickle_commit_store_aborts_skipped_reservations`. `tests/v1/distributed` plus `test_engine_driven_transfer.py` on GPU show the same single pre-existing failure on dev and on this branch, with the 4 new tests passing. On CPU, `tests/v1/distributed`, `tests/v1/multiprocess` and the adapter tests give identical failure sets on dev and on this branch.

### CODE REVIEW

File: tests/v1/distributed/test_abort_write.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+
+        def on_l2_keys_deleted(self, keys):
+            pass
+
+    storage_manager._l1_manager.register_listener(_Listener())
+    keys = [make_object_key(10)]
+    storage_manager.reserve_write(keys, basic_layout, "new")
+    storage_manager.abort_write(keys)
+
+    assert events == [] or all(not e for e in events)
```
[the remark this answers is not in the document]
Comment (divyvasal) [reply tests/v1/distributed/test_abort_write.py]: Done — both listener-notification and state-rejection tests now build an `L1Manager` directly (new `l1_manager` fixture) instead of reaching into `StorageManager._l1_manager`. The listener test also gained a positive assert that `on_l1_keys_deleted_by_manager` fires for aborted keys. Re-ran the suite on a CUDA box: 40 passed, 2 skipped.

File: tests/v1/distributed/test_abort_write.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    storage_manager.reserve_write(keys, basic_layout, "new")
+    storage_manager.finish_write(keys)
+
+    result = storage_manager._l1_manager.abort_write(keys)
+    assert result[keys[0]] == L1Error.KEY_IN_WRONG_STATE
+
+    missing = [make_object_key(21)]
+    result = storage_manager._l1_manager.abort_write(missing)
+    assert result[missing[0]] == L1Error.KEY_NOT_EXIST
+
```
[the remark this answers is not in the document]
Comment (divyvasal) [reply tests/v1/distributed/test_abort_write.py]: Done — same fix as above: the test now instantiates `L1Manager` directly via the `l1_manager` fixture; no private member access remains in the file.

**Comment by divyvasal** [comment]
@ApostaC @maobaolong @chunxiaozheng — review pass appreciated when you have a cycle.

Small, self-contained fix with a production incident behind it: skipped write reservations in the engine_driven pickle path leak L1 write locks until every reserve fails (we hit 604 leaked locks / 91.5% L1 in ~30 min), and two neighboring sites publish never-written objects via `finish_write`. Adds an `abort_write` primitive + uses it at all three sites. DCO green, Gemini review addressed, 40 tests green on CUDA. Companion issue: #4069; surfaced while productionizing #3892.

**Comment by divyvasal** [comment]
@ApostaC when you get a chance — green + mergeable, a small correctness fix (abort skipped write reservations instead of committing them). No rush, just flagging it's ready for review.

**Comment by divyvasal** [comment]
Rebased on current dev and re-verified on GPU (details in the description). Still relevant: all three sites are unchanged on dev. Not stale, ready for review.
